src/Lean/Compiler/IR/LLVMBindings.lean

```diff
@@ -0,0 +1,316 @@
+/-
+Copyright (c) 2022 Microsoft Corporation. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Siddharth Bhat
+-/
+
+namespace LLVM
+/-!
+This file defines bindings for LLVM. The main mechanisms
+are to:
+- Mirror the values of C enumerations.
+- Create opaque values for LLVM's pointer based API
+- Write bindings that are `IO` based, which mutate the underlying in-memory
+  data structures to build IR.
+-/
+
+-- https://github.com/llvm/llvm-project/blob/c3e073bcbdc523b0f758d44a89a6333e38bff863/llvm/include/llvm-c/TargetMachine.h#L64
+structure CodegenFileType where
+  private mk :: val : UInt64
+
+def CodegenFileType.AssemblyFile : CodegenFileType := { val := 0 }
+def CodegenFileType.ObjectFile : CodegenFileType := { val := 1 }
+
+-- https://github.com/llvm/llvm-project/blob/c3e073bcbdc523b0f758d44a89a6333e38bff863/llvm/include/llvm-c/Core.h#L290
+structure IntPredicate where
+  private mk :: val : UInt64
+
+def IntPredicate.EQ : IntPredicate := { val := 32 }
+def IntPredicate.NE : IntPredicate := { val := IntPredicate.EQ.val + 1 }
+def IntPredicate.UGT : IntPredicate := { val := IntPredicate.NE.val + 1 }
+
+structure BasicBlock (ctx : Context)  where
+  private mk :: ptr : USize
+  deriving Inhabited
+
+structure Builder (ctx : Context) where
+  private mk :: ptr : USize
+  deriving Inhabited
+
+structure Context where
+  private mk :: ptr : USize
+  deriving Inhabited
+
+structure LLVMType (ctx : Context) where
+  private mk :: ptr : USize
+  deriving Inhabited
+
+structure MemoryBuffer (ctx : Context) where
+  private mk :: ptr : USize
+  deriving Inhabited
+
+structure Module (ctx : Context) where
+  private mk :: ptr : USize
+  deriving Inhabited
+
+structure Target (ctx : Context) where
+  private mk :: ptr : USize
+  deriving Inhabited
+
+structure TargetMachine (ctx : Context) where
+  private mk :: ptr : USize
+  deriving Inhabited
+
+structure Value (ctx : Context) where
+  private mk :: ptr : USize
+  deriving Inhabited
+
+@[extern "lean_llvm_create_context"]
+opaque createContext : BaseIO (Context)
+
+@[extern "lean_llvm_create_module"]
+opaque createModule (ctx : @&Context) (name : @&String) : BaseIO (Module ctx)
+
+@[extern "lean_llvm_module_to_string"]
+opaque moduleToString (m : @&Module ctx) : BaseIO String
+
+@[extern "lean_llvm_write_bitcode_to_file"]
+opaque writeBitcodeToFile (m : @&Module ctx) (path : @&String) : BaseIO Unit
+
+@[extern "lean_llvm_add_function"]
+opaque addFunction (m : @&Module ctx) (name : @&String) (type : @&LLVMType ctx) : BaseIO (Value ctx)
```

Comment:
```suggestion
opaque addFunction (m : Module ctx) (name : @& String) (type : LLVMType ctx) : BaseIO (Value ctx)
```
All of the borrow annotations here are moot btw, since `Module` etc. are all value types (`USize`).